File: sdk/swift/DevilChainSDK.swift
// DevilChain Swift/iOS SDK
import Foundation

public class DevilChainSDK {
    private let baseURL: String
    public init(baseURL: String = "http://localhost:8545") { self.baseURL = baseURL }

    private func get(_ path: String) async throws -> [String: Any] {
        guard let url = URL(string: baseURL + path) else { throw URLError(.badURL) }
        let (data, _) = try await URLSession.shared.data(from: url)
        return try JSONSerialization.jsonObject(with: data) as? [String: Any] ?? [:]
    }
    private func post(_ path: String, body: [String: Any]) async throws -> [String: Any] {
        guard let url = URL(string: baseURL + path) else { throw URLError(.badURL) }
        var req = URLRequest(url: url)
        req.httpMethod = "POST"; req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try JSONSerialization.data(withJSONObject: body)
        let (data, _) = try await URLSession.shared.data(for: req)
        return try JSONSerialization.jsonObject(with: data) as? [String: Any] ?? [:]
    }
    public func getStatus() async throws -> [String: Any] { try await get("/api/status") }
    public func getLatestBlock() async throws -> [String: Any] { try await get("/api/block/latest") }
    public func getWallet(_ addr: String) async throws -> [String: Any] { try await get("/api/wallet/\(addr)") }
    public func sendTransaction(from: String, to: String, amount: Double, gasFee: Double, signature: String) async throws -> [String: Any] {
        try await post("/api/send", body: ["from": from, "to": to, "amount": amount, "gas_fee": gasFee, "signature": signature])
    }
    public func stake(address: String, amount: Double, signature: String) async throws -> [String: Any] {
        try await post("/api/stake", body: ["address": address, "amount": amount, "signature": signature])
    }
}

File: sdk/kotlin/DevilChainSDK.kt
package network.devilchain.sdk

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL

class DevilChainSDK(private val baseUrl: String = "http://localhost:8545") {
    private suspend fun get(path: String): JSONObject = withContext(Dispatchers.IO) {
        val conn = URL("$baseUrl$path").openConnection() as HttpURLConnection
        conn.requestMethod = "GET"; conn.connect()
        JSONObject(conn.inputStream.bufferedReader().readText())
    }
    private suspend fun post(path: String, body: JSONObject): JSONObject = withContext(Dispatchers.IO) {
        val conn = URL("$baseUrl$path").openConnection() as HttpURLConnection
        conn.requestMethod = "POST"; conn.setRequestProperty("Content-Type", "application/json")
        conn.doOutput = true; conn.outputStream.write(body.toString().toByteArray()); conn.connect()
        JSONObject(conn.inputStream.bufferedReader().readText())
    }
    suspend fun getStatus() = get("/api/status")
    suspend fun getLatestBlock() = get("/api/block/latest")
    suspend fun getBlock(h: Int) = get("/api/block/$h")
    suspend fun getWallet(addr: String) = get("/api/wallet/$addr")
    suspend fun getValidators() = get("/api/validators")
    suspend fun getDaoProposals() = get("/api/dao/proposals")
    suspend fun sendTransaction(from: String, to: String, amount: Double, gasFee: Double, sig: String) =
        post("/api/send", JSONObject().apply { put("from",from); put("to",to); put("amount",amount); put("gas_fee",gasFee); put("signature",sig) })
    suspend fun stake(addr: String, amount: Double, sig: String) =
        post("/api/stake", JSONObject().apply { put("address",addr); put("amount",amount); put("signature",sig) })
}
